chain/epoch-manager/src/validator_selection.rs

```diff
@@ -14,6 +14,97 @@ use std::cmp::{self, Ordering};
+    block_producers: Vec<ValidatorStake>,
+    /// Validators which are assigned to validate chunks.
+    chunk_validators: Vec<ValidatorStake>,
+    /// Stake threshold to become a validator.
+    threshold: Balance,
+}
+
+/// Helper struct which is a result of assigning chunk producers to shards.
+struct ChunkProducersAssignment {
+    /// List of all validators in the epoch.
+    /// Note that it doesn't belong here, but in the legacy code it is computed
+    /// together with chunk producers assignment.
+    all_validators: Vec<ValidatorStake>,
+    /// Maps validator account names to local indices throughout the epoch.
+    validator_to_index: HashMap<AccountId, ValidatorId>,
+    /// Maps chunk producers to shards, where i-th list contains local indices
+    /// of validators producing chunks for i-th shard.
+    chunk_producers_settlement: Vec<Vec<ValidatorId>>,
+}
+
+/// Selects validator roles for the given proposals.
+fn select_validators_from_proposals(
+    epoch_config: &EpochConfig,
+    proposals: HashMap<AccountId, ValidatorStake>,
+    next_version: ProtocolVersion,
+) -> ValidatorRoles {
+    let shard_ids: Vec<_> = epoch_config.shard_layout.shard_ids().collect();
+    let min_stake_ratio = {
+        let rational = epoch_config.validator_selection_config.minimum_stake_ratio;
+        Ratio::new(*rational.numer() as u128, *rational.denom() as u128)
+    };
+
+    let mut chunk_producer_proposals = order_proposals(proposals.values().cloned());
+    let (chunk_producers, cp_stake_threshold) = select_chunk_producers(
+        &mut chunk_producer_proposals,
+        epoch_config.validator_selection_config.num_chunk_producer_seats as usize,
+        min_stake_ratio,
+        shard_ids.len() as NumShards,
+        next_version,
+    );
+
+    let mut block_producer_proposals = order_proposals(proposals.values().cloned());
+    let (block_producers, bp_stake_threshold) = select_block_producers(
+        &mut block_producer_proposals,
+        epoch_config.num_block_producer_seats as usize,
+        min_stake_ratio,
+        next_version,
+    );
+
+    let mut chunk_validator_proposals = order_proposals(proposals.into_values());
+    let (chunk_validators, cv_stake_threshold) = select_validators(
+        &mut chunk_validator_proposals,
+        epoch_config.validator_selection_config.num_chunk_validator_seats as usize,
+        min_stake_ratio,
+        next_version,
+    );
+
+    // Note that if there are too few validators and too many shards,
+    // assigning chunk producers to shards is more aggressive, so it
+    // is not enough to iterate over chunk validators.
+    // So unfortunately we have to look over all roles to get unselected
+    // proposals.
+    // TODO: getting unselected proposals must be simpler. For example,
+    // we can track all roles assign to each validator in some structure
+    // and then return all validators which don't have any role.
+    let max_validators_for_role =
```

Comment:
I do not get this part. What does unselected proposals represent? I first thought these are the validators that we are unable to find a role but you are not assigning the entire set of block or chunk producers/validators to that set. Do I misunderstand the semantics of unselected proposals?

Reply:
I stated the purpose in comment:
```
ValidatorRoles {
    /// Proposals which were not given any role.
   unselected_proposals: BinaryHeap<OrderedValidatorStake>,
```
So "validators that we are unable to find a role" is right.

> you are not assigning the entire set of block or chunk producers/validators to that set.

Why would I need it? The goal is to get validators without roles, not the opposite.
One way to do it is to take one of `chunk_validator_proposals/...` because all proposals are selected by stake. After `select_validators/...` is called, only unselected proposals are left in this set. Selected proposals go to `chunk_validators`.

Please note that this is how legacy selection logic currently works and I want to make changes iteratively to make it better with time.

Reply:
I agree with @tayfunelmas that this is very confusing to read. How about :

a) Rather than mutating the `x_proposals` you pass it into the `select_x` and have the unselected ones returned? 
```
  let ordered_proposals = order_proposals(proposals.values().cloned());
  let (chunk_producers, not_chunk_producers, cp_stake_threshold) = select_chunk_producers(
    ordered_proposals,
    ...
  );
```

or 

b) To find the unselected ones just iterate over all proposals and check which ones are not in bp, cp, val sets.  
```
for proposal in ordered_proposals {
  if chunk_producers.contains(proposal) continue;
  if block_producers.contains(proposal) continue;
  if chunk_validators.contains(proposal) continue;
  unselected.insert(proposal);
}
```